app/policies/result_policy.rb

```diff
@@ -49,6 +50,19 @@ def review?
     )
   end
 
+  def update_mark?
+    assigned_to_criterion = true
+    assignment = record.grouping.assignment
+    if assignment.assign_graders_to_criteria && role.ta?
+      assigned_to_criterion = role.criterion_ta_associations.where(criterion_id: criterion_id).present?
+    end
+
+    check?(:manage_submissions?, role) || (check?(:assigned_grader?, record.grouping) &&
```

Comment:
Overall you can do a bit more code reuse by calling `check(:review?` as a helper. Then have a separate check for whether `role.criterion_ta_associations.where(criterion_id: criterion_id).present?` is true or not.